buildtest/buildsystem/base.py

```diff
@@ -640,3 +644,274 @@ def _build_testcontent(self):
+            intel = IntelCompiler()
+            self.cc, self.cxx, self.fc = intel.get_compilers()
+
+        # Get PGI Compiler
+        elif self.compiler_recipe.get("pgi"):
+            self.compiler_name = "pgi"
+            pgi = PGICompiler()
+            self.cc, self.cxx, self.fc = pgi.get_compilers()
+
+        # Get Cray Compiler
+        elif self.compiler_recipe.get("cray"):
+            self.compiler_name = "cray"
+            cray = CrayCompiler()
+            self.cc, self.cxx, self.fc = cray.get_compilers()
+
+        # Set ldflags, cppflags, cflags, cxxflags, fflags if defined in Buildspec
+        self.ldflags = self.compiler_recipe[self.compiler_name].get("ldflags")
+        self.cppflags = self.compiler_recipe[self.compiler_name].get("cppflags")
+        self.cflags = self.compiler_recipe[self.compiler_name].get("cflags")
+        self.fflags = self.compiler_recipe[self.compiler_name].get("fflags")
+        self.cxxflags = self.compiler_recipe[self.compiler_name].get("cxxflags")
+
+        cmd = self.generate_compile_cmd()
+
+        # every test starts with shebang line
+        lines = [self.shebang]
+
+        # if 'module' defined in Buildspec add modules to test
+        if self.recipe.get("module"):
+            for module in self.recipe.get("module"):
+                lines.append(module)
+
+        # add compile command to test
+        lines.append(" ".join(cmd))
+
+        # build run command that launches generated executable. Format: <exec> <args>
+        # if 'exec_args' is not defined it will return 'None' which means no arguments to be passed when running executable
+        run = [
+            self.executable,
+            self.compiler_recipe[self.compiler_name].get("exec_args"),
+        ]
+
+        # remove 'None' in case 'exec_args' is not defined in Buildspec
+        run = list(filter(None, run))
+
+        # add run line to test
+        lines.append(" ".join(run))
+        return lines
+
+    def detect_lang(self, sourcefile):
+        """This method will return the Programming Language based by looking up  file extension of source file."""
+
+        ext = os.path.splitext(sourcefile)[1]
+
+        # if ext not in self.lang_ext_table then raise an error. This table consist of all file extensions that map to a Programming Language
+        if ext not in self.lang_ext_table:
+            raise BuildTestError(
+                f"Unable to detect Program Language based on extension: {ext} in file {sourcefile}"
+            )
+        # Set Programming Language based on ext. Programming Language could be (C, C++, Fortran)
+        lang = self.lang_ext_table[ext]
+        return lang
+
+    def generate_compile_cmd(self):
+        """This method generates the compilation line and returns the output as a list. The compilation line depends
+           on the the language detected that is stored in variable ``self.lang``.
+        """
+        cmd = []
+
+        # Generate C compilation line
```

Comment:
Another example here of if else elif that will be addressed by separation into the appropriate classes. I shouldn't need to edit the Abstract Compiler Base if I want to tweak the FortranCompiler, for example. This is what I'd have to do now.

Reply:
So the way C, C++, Fortran compilation works for single program is as follows, this is true for any compiler to the best of my knowledge.

```
$cc $CPPFLAGS $CFLAGS -o $EXEC $SOURCE $LDFLAGS
$cxx $CPPFLAGS $CXXFLAGS -o $EXEC $SOURCE $LDFLAGS
$fc $CPPFLAGS $FFLAGS -o $EXEC $SOURCE $LDFLAGS
```

The $cc is your C compiler with CFLAGS as option to C compiler, the same is true with Fortran (FFLAGS) and C++ (CXXFLAGS). 
That's how Makefiles work if you take look at this example, you just need to change cc and CFLAGS to map to the compiler you want but the pattern for compilation remains the same.
https://www.cs.swarthmore.edu/~newhall/unixhelp/howto_makefiles.html

That's why i implemented this in Base class. The one issue we have right now is CompilerBuilder is base class of BuilderBase in order to inherit the build logic.  The compilation pattern can change if user wants to build source into object files (.o) and link each object into executable. 

Here is some example https://www3.ntu.edu.sg/home/ehchua/programming/cpp/gcc_make.html for compilation

Reply:
Thank you for your explanation of how compilers work - I'm aware :) This is different from the refactor needed to clean up the base class.